src/transformers/models/realm/modeling_realm.py

```diff
@@ -0,0 +1,1309 @@
+        candidate_inputs_embeds (:obj:`torch.FloatTensor` of shape :obj:`(batch_size * num_candidates, sequence_length, hidden_size)`, `optional`):
+            Optionally, instead of passing :obj:`candidate_input_ids` you can choose to directly pass an embedded
+            representation. This is useful if you want more control over how to convert `candidate_input_ids` indices
+            into associated vectors than the model's internal embedding lookup matrix.
+
+        Returns:
+        """
+
+        return_dict = return_dict if return_dict is not None else self.config.use_return_dict
+
+        if not (
+            any((input_ids is not None, inputs_embeds is not None))
+            and any((candidate_input_ids is not None, candidate_inputs_embeds is not None))
+        ):
+            raise ValueError("You have to specify both inputs and candidate inputs")
+
+        query_outputs = self.query_embedder(
+            input_ids,
+            attention_mask=attention_mask,
+            token_type_ids=token_type_ids,
+            position_ids=position_ids,
+            head_mask=head_mask,
+            inputs_embeds=inputs_embeds,
+            output_attentions=output_attentions,
+            output_hidden_states=output_hidden_states,
+            return_dict=return_dict,
+        )
+
+        # [batch_size * num_candidates, candidate_seq_len]
+        (flattened_input_ids, flattened_attention_mask, flattened_token_type_ids) = self._flatten_inputs(
+            candidate_input_ids, candidate_attention_mask, candidate_token_type_ids
+        )
+
+        candidate_outputs = self.embedder(
+            flattened_input_ids,
+            attention_mask=flattened_attention_mask,
+            token_type_ids=flattened_token_type_ids,
+            position_ids=position_ids,
+            head_mask=head_mask,
+            inputs_embeds=candidate_inputs_embeds,
+            output_attentions=output_attentions,
+            output_hidden_states=output_hidden_states,
+            return_dict=return_dict,
+        )
+
+        # [batch_size, retriever_proj_size]
+        query_score = query_outputs[0]
+        # [batch_size * num_candidates, retriever_proj_size]
+        candidate_score = candidate_outputs[0]
+        # [batch_size, num_candidates, retriever_proj_size]
+        candidate_score = candidate_score.view(-1, self.config.num_candidates, self.config.retriever_proj_size)
+        # [batch_size, num_candidates]
+        relevance_score = torch.einsum("BD,BND->BN", query_score, candidate_score)
+
+        if not return_dict:
+            return relevance_score, query_score, candidate_score
+
+        return RealmRetrieverOutput(
+            relevance_score=relevance_score, query_score=query_score, candidate_score=candidate_score
+        )
+
+
+@add_start_docstrings(
+    "The encoder of REALM outputting masked language model logits and marginal log-likelihood loss.",
+    REALM_START_DOCSTRING,
+)
+class RealmEncoder(RealmPreTrainedModel):
```

Comment:
Maybe we can rename this class here - it seems like this model is trained using the MLM loss -> could we name the model maybe `RealmForMaskedLM`?

Reply:
Good point! It's called "Knowledge-augmented Encoder" in the paper no? Could we maybe call it `RealmKnowledgeAugEncoder` then?

Reply:
Sounds great! This name is better I believe.